File: garnet/examples/fidl/echo_client_go/echo_client.go
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// +build !build_with_native_toolchain

package main

import (
	"context"
	"flag"
	"fmt"
	"syscall/zx"
	"syscall/zx/fdio"

	"go.fuchsia.dev/fuchsia/src/lib/component"

	"fidl/fidl/examples/echo"
	"fidl/fuchsia/io"
	"fidl/fuchsia/sys"
)

func main() {
	serverURL := flag.String("server", "fuchsia-pkg://fuchsia.com/echo_server_go#meta/echo_server_go.cmx", "server URL")
	msg := flag.String("m", "Hello, Go World", "message")

	flag.Parse()

	ctx := component.NewContextFromStartupInfo()

	directoryReq, directoryInterface, err := io.NewDirectoryWithCtxInterfaceRequest()
	if err != nil {
		panic(err)
	}
	launchInfo := sys.LaunchInfo{
		Url:              *serverURL,
		DirectoryRequest: directoryReq.Channel,
	}

	componentControllerReq, _, err := sys.NewComponentControllerWithCtxInterfaceRequest()
	if err != nil {
		panic(err)
	}
	if err := ctx.Launcher().CreateComponent(context.Background(), launchInfo, componentControllerReq); err != nil {
		panic(err)
	}

	echoReq, echoInterface, err := echo.NewEchoWithCtxInterfaceRequest()
	if err != nil {
		panic(err)
	}
	if err := fdio.ServiceConnectAt(zx.Handle(directoryInterface.Channel), echoReq.Name(), zx.Handle(echoReq.Channel)); err != nil {
		panic(err)
	}

	response, err := echoInterface.EchoString(context.Background(), msg)
	if err != nil {
		panic(err)
	}
	fmt.Println("client:", *response)
}

File: examples/fidl/legacy/echo_client_cpp_synchronous/echo_client.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fidl/examples/echo/cpp/fidl.h>
#include <lib/async-loop/cpp/loop.h>
#include <lib/async-loop/default.h>
#include <lib/async/default.h>
#include <lib/zx/process.h>
#include <zircon/processargs.h>

#include "echo_client_app.h"

int main(int argc, const char** argv) {
  std::string server_url = "fuchsia-pkg://fuchsia.com/echo_server_cpp#meta/echo_server_cpp.cmx";
  std::string msg = "hello synchronous world";

  for (int i = 1; i < argc - 1; ++i) {
    if (!strcmp("--server", argv[i])) {
      server_url = argv[++i];
    } else if (!strcmp("-m", argv[i])) {
      msg = argv[++i];
    }
  }

  async::Loop loop(&kAsyncLoopConfigAttachToCurrentThread);

  echo::EchoClientApp app;
  app.Start(server_url);

  ::fidl::StringPtr result;
  app.echo_sync()->EchoString(msg, &result);
  if (result) {
    printf("***** Sync Response: %s\n", result->data());
  } else {
    printf("***** Sync Response: none\n");
  }
  return 0;
}
